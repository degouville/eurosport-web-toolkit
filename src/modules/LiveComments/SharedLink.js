import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import Cards from '../../components/Card';

const SharedLink = ({ sharedlink, labelPlayButton }) => {
  const image = get(sharedlink, ['pictureurl'], '');

  const sharedLinkProps = {
    title: get(sharedlink, ['title'], ''),
    labelPlayButton: get(sharedlink, ['link', 'type'], '') === 2 ? labelPlayButton : '',
    image: image ? `${image}?w=640` : '',
    link: get(sharedlink, ['url'], ''),
  };

  return <Cards.CardSmall {...sharedLinkProps} target="_blank" />;
};

export const SharedLinkShape = PropTypes.shape({
  title: PropTypes.string,
  url: PropTypes.string,
  link: PropTypes.shape({
    type: PropTypes.number,
  }),
  pictureurl: PropTypes.string,
});

SharedLink.propTypes = {
  sharedlink: SharedLinkShape.isRequired,
  labelPlayButton: PropTypes.string.isRequired,
};

export default SharedLink;
