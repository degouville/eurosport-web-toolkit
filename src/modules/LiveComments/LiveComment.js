import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';
import get from 'lodash/get';
import { medium, large, wide } from '../../breakpoints';
import { jungleMist, halfBaked, turquoiseBlue, coreLightMinus1 } from '../../colors';
import { fontAlphaHeadline } from '../../typography';

import Cards from '../../components/Card';
import SharedLink, { SharedLinkShape } from './SharedLink';

export const StyledComment = styled.div`
  margin-top: 50px;
  display: flex;
`;
const StyledCommentLeft = styled.div`
  flex-shrink: 0;
  text-align: center;
  color: ${turquoiseBlue};
  font-size: 10px;
  font-weight: bold;
  width: 32px;
  margin-right: 20px;

  ${large(css`
    font-size: 12px;
    width: 42px;
  `)}

  ${wide(css`
    width: 62px;
  `)}
`;
const StyledCommentRight = styled.div`
  flex: 1;
  overflow: hidden;
  color: ${coreLightMinus1};
`;

export const StyledPlainHtml = styled.div`
  * {
    font-size: 14px;
    line-height: 20px;
    ${fontAlphaHeadline};

    ${large(css`
      font-size: 16px;
      line-height: 22px;
    `)}
  }

  a {
    color: ${halfBaked};

    &:hover {
      color: ${jungleMist};
    }
  }

  p {
    margin: 0;
  }
  b {
    font-weight: bold;
  }
  i {
    font-style: italic;
  }
  u {
    text-decoration: underline;
  }
`;
export const StyledMarker = styled.div`
  ${fontAlphaHeadline};
  margin-bottom: 5px;
`;
const StyledPictogram = styled.img`
  width: 17px;
  max-height: 17px;
`;

export const StyledTwitterCard = styled(Cards.Twitter)`
  margin-top: 25px;
`;

const StyledCardWrapper = styled.div`
  margin-top: 25px;
  max-width: 308px;
  text-decoration: none;

  ${medium(css`
    max-width: 388px;
  `)}

  ${large(css`
    max-width: 342px;
  `)}
  
  ${wide(css`
    max-width: 308px;
  `)}
  
  img {
    max-width: 100%;
    border-radius: 4px;
  }
`;

const LiveComment = ({ liveComment, labelPlayButton }) => {
  const iconUrl = get(liveComment, ['icon', 'svg'], null);
  const pictureUrl = get(liveComment, ['picture', 'format', 'url'], null);
  const shouldDisplayComment = !!liveComment.html || !!liveComment.marker || !!iconUrl || !!liveComment.tweet;

  return shouldDisplayComment ? (
    <StyledComment>
      <StyledCommentLeft>
        {!!liveComment.marker && <StyledMarker>{liveComment.marker}</StyledMarker>}
        {iconUrl && <StyledPictogram src={iconUrl} />}
      </StyledCommentLeft>
      <StyledCommentRight>
        {liveComment.html && <StyledPlainHtml dangerouslySetInnerHTML={{ __html: liveComment.html }} />}
        {pictureUrl && (
          <StyledCardWrapper>
            {/* eslint-disable-next-line jsx-a11y/alt-text */}
            <img src={`${pictureUrl}?w=640`} />
          </StyledCardWrapper>
        )}
        {liveComment.tweet && <StyledTwitterCard tweetUrl={get(liveComment, ['tweet', 'url'], '')} />}
        {liveComment.sharedlink && (
          <StyledCardWrapper>
            <SharedLink sharedlink={liveComment.sharedlink} labelPlayButton={labelPlayButton} />
          </StyledCardWrapper>
        )}
      </StyledCommentRight>
    </StyledComment>
  ) : null;
};

LiveComment.defaultProps = {
  labelPlayButton: 'watch',
};

export const LiveCommentPropTypeShape = PropTypes.shape({
  id: PropTypes.number,
  icon: PropTypes.shape({
    svg: PropTypes.string,
  }),
  sharedlink: SharedLinkShape,
  marker: PropTypes.string,
  html: PropTypes.string,
  picture: PropTypes.shape({
    format: PropTypes.shape({
      url: PropTypes.string,
    }),
  }),
  tweet: PropTypes.shape({
    url: PropTypes.string,
  }),
});

LiveComment.propTypes = {
  liveComment: LiveCommentPropTypeShape.isRequired,
  labelPlayButton: PropTypes.string,
};

export default LiveComment;
