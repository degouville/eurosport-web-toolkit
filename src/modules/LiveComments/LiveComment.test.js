import React from 'react';
import { shallow } from 'enzyme/build';
import SharedLink from './SharedLink';
import LiveComment, { StyledMarker, StyledPlainHtml, StyledTwitterCard } from './LiveComment';

describe('LiveComment', () => {
  const liveComment = {
    html: '<b>comment html text</b>',
    marker: 'comment marker text',
    tweet: {
      url: 'tweetUrl',
    },
    picture: {
      format: {
        url: 'picture.format.url',
      },
    },
    sharedlink: {
      title: 'shared link title',
      image: 'shared.link.image',
      link: {
        type: 2,
      },
    },
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<LiveComment liveComment={liveComment} labelPlayButton="play button" />);
  });

  it('matches snapshot with all valid props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('render cases', () => {
    it('null if html, marker, pictureurl and tweet are missing', () => {
      wrapper.setProps({
        liveComment: {},
      });

      expect(wrapper.html()).toBe(null);
    });

    it('skips marker if one is missing', () => {
      wrapper.setProps({
        liveComment: {
          ...liveComment,
          marker: null,
        },
      });
      expect(wrapper.find(StyledMarker)).toHaveLength(0);
    });

    it('skips html if one is missing', () => {
      wrapper.setProps({
        liveComment: {
          ...liveComment,
          html: null,
        },
      });
      expect(wrapper.find(StyledPlainHtml)).toHaveLength(0);
    });

    it('skips picture if one is missing', () => {
      wrapper.setProps({
        liveComment: {
          ...liveComment,
          picture: {
            format: {},
          },
        },
      });
      expect(wrapper.find('img')).toHaveLength(0);
    });

    it('skips tweet if one is missing', () => {
      wrapper.setProps({
        liveComment: {
          ...liveComment,
          tweet: null,
        },
      });
      expect(wrapper.find(StyledTwitterCard)).toHaveLength(0);
    });

    it('skips sharedlink if one is missing', () => {
      wrapper.setProps({
        liveComment: {
          ...liveComment,
          sharedlink: null,
        },
      });
      expect(wrapper.find(SharedLink)).toHaveLength(0);
    });
  });
});
