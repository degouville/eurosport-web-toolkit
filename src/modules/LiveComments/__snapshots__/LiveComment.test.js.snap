// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveComment matches snapshot with all valid props 1`] = `
<Styled(div)>
  <Styled(div)>
    <Styled(div)>
      comment marker text
    </Styled(div)>
  </Styled(div)>
  <Styled(div)>
    <Styled(div)
      dangerouslySetInnerHTML={
        Object {
          "__html": "<b>comment html text</b>",
        }
      }
    />
    <Styled(div)>
      <img
        src="picture.format.url?w=640"
      />
    </Styled(div)>
    <Styled(Twitter)
      tweetUrl="tweetUrl"
    />
    <Styled(div)>
      <SharedLink
        labelPlayButton="play button"
        sharedlink={
          Object {
            "image": "shared.link.image",
            "link": Object {
              "type": 2,
            },
            "title": "shared link title",
          }
        }
      />
    </Styled(div)>
  </Styled(div)>
</Styled(div)>
`;
